matchms/similarity/ModifiedCosine.py

```diff
@@ -0,0 +1,102 @@
+import numba
+import numpy
+from matchms.typing import SpectrumType
+
+
+class ModifiedCosine:
+    """Calculate modified cosine score between spectra.
+
+    This score is calculated based on matches between peaks of two spectra and
+    also includes a shift of peak positions by the difference in precursor mass.
+
+    Args:
+    ----
+    tolerance: float
+        Peaks will be considered a match when <= tolerance appart. Default is 0.1.
+    """
+    def __init__(self, tolerance=0.1):
+        self.tolerance = tolerance
+
+    def __call__(self, spectrum1: SpectrumType, spectrum2: SpectrumType) -> float:
+        """Calculate modified cosine score between two spectra.
+        Args:
+        ----
+        spectrum1: SpectrumType
+            Input spectrum 1.
+        spectrum2: SpectrumType
+            Input spectrum 2.
+        """
+        def get_peaks_arrays():
+            """Get peaks mz and intensities as numpy array."""
+            spec1 = numpy.vstack((spectrum1.peaks.mz, spectrum1.peaks.intensities)).T
+            spec2 = numpy.vstack((spectrum2.peaks.mz, spectrum2.peaks.intensities)).T
+            assert max(spec1[:, 1]) <= 1, ("Input spectrum1 is not normalized. ",
+                                           "Apply 'normalize_intensities' filter first.")
+            assert max(spec2[:, 1]) <= 1, ("Input spectrum2 is not normalized. ",
+                                           "Apply 'normalize_intensities' filter first.")
+            return spec1, spec2
+
+        def get_matching_pairs():
+            """Find all pairs of peaks that match within the given tolerance."""
+            zero_pairs = find_pairs_numba(spec1, spec2, self.tolerance, shift=0.0)
+            if spectrum1.get("precursor_mz") and spectrum2.get("precursor_mz"):
+                mass_shift = spectrum1.get("precursor_mz") - spectrum2.get("precursor_mz")
+                nonzero_pairs = find_pairs_numba(spec1, spec2, self.tolerance, shift=mass_shift)
+            else:
+                print("Precursor_mz missing. Calculate cosine score instead.",
+                      "Consider applying 'add_precursor_mz' filter first.")
+                nonzero_pairs = []
```

Comment:
A print statement can be missed, please change to assertion so code raises exception
```suggestion
            message = "Precursor_mz missing. Apply 'add_precursor_mz' filter first."
            assert spectrum1.get("precursor_mz") and spectrum2.get("precursor_mz"), message
            mass_shift = spectrum1.get("precursor_mz") - spectrum2.get("precursor_mz")
            nonzero_pairs = find_pairs_numba(spec1, spec2, self.tolerance, shift=mass_shift)
```

Reply:
That's better, thanks!